scheduler/scheduler_shuffle.go

```diff
@@ -0,0 +1,64 @@
+package scheduler
+
+import (
+	"context"
+	"math/rand"
+	"sync"
+
+	"github.com/cloudquery/plugin-sdk/v4/schema"
+)
+
+func (s *syncClient) syncShuffle(ctx context.Context, resolvedResources chan<- *schema.Resource) {
+	// we have this because plugins can return sometimes clients in a random way which will cause
+	// differences between this run and the next one.
+	preInitialisedClients := make([][]schema.ClientMeta, len(s.tables))
+	for i, table := range s.tables {
+		clients := []schema.ClientMeta{s.client}
+		if table.Multiplex != nil {
+			clients = table.Multiplex(s.client)
+		}
+		preInitialisedClients[i] = clients
+		// we do this here to avoid locks so we initial the metrics structure once in the main goroutines
+		// and then we can just read from it in the other goroutines concurrently given we are not writing to it.
+		s.metrics.initWithClients(table, clients)
+	}
+
+	// first interleave the tables like in round-robin
+	tableClients := roundRobinInterleave(s.tables, preInitialisedClients)
+
+	// then shuffle the tableClients to randomize the order in which they are retrieved
+	shuffle(tableClients)
+
+	var wg sync.WaitGroup
+	for _, tc := range tableClients {
+		table := tc.table
+		cl := tc.client
+		if err := s.scheduler.tableSems[0].Acquire(ctx, 1); err != nil {
+			// This means context was cancelled
+			wg.Wait()
+			return
+		}
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			defer s.scheduler.tableSems[0].Release(1)
+			// not checking for error here as nothing much to do.
+			// the error is logged and this happens when context is cancelled
+			// This currently uses the DFS algorithm to resolve the tables, but this
+			// may change in the future.
+			s.resolveTableDfs(ctx, table, cl, nil, resolvedResources, 1)
+		}()
+	}
+
+	// Wait for all the worker goroutines to finish
+	wg.Wait()
+}
+
+func shuffle(tableClients []tableClient) {
+	// use a fixed seed so that runs with the same tables and clients perform similarly across syncs
+	r := rand.New(rand.NewSource(99))
```

Comment:
Doesn't this mean we always randomize it the same way 🙃 ? Ideally this would be configurable but maybe we can use something from the spec so if the initial shuffle is bad users can modify it. Maybe hash on table/client IDs or something

Reply:
> Doesn't this mean we always randomize it the same way 

First the obligatory https://xkcd.com/221/

>  Maybe hash on table/client IDs or something

Yeah we can do that, so then changing the ordering of tables in your config will cause a different shuffle.

Reply:
I don't think it was really necessary, the odds of getting a bad shuffle (at either end of the DFS->Round Robin spectrum) if there are enough table-clients is astronomically small, but in [27c4e22](https://github.com/cloudquery/plugin-sdk/pull/1218/commits/27c4e22fc4dd2db79b2e7ede35ecd1c135afe28c) I added the ability to control it a little bit by changing the ordering of the tables in the config.